Add HAL CRC test and header file

mprse opened the pull request:
### Description

Add CRC HAL API test and test header file.

Since the testing requirements were not provided in PR https://github.com/ARMmbed/mbed-os/pull/6708 I allowed myself to define them here.

@scartmell-arm Please review the test and testing requirements and update if something is wrong or should be additionally tested.

### Status

READY

### Related PRs

https://github.com/ARMmbed/mbed-os/pull/6708 - provides CRC HAL API definition and example driver for K64F. Needs to be merged before this one.

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] New target
    [X] Feature
    [ ] Breaking change

mprse committed:
Add testing requirements to HAL CRC header file.

mprse committed:
Add HAL CRC test and test header file.

cmonr commented:
@scartmell-arm hal/crc_api.h exists in both this PR and https://github.com/ARMmbed/mbed-os/pull/6708. Was that the intention, or is this PR only suppose to contain tests?

ghost commented:
> @scartmell-arm hal/crc_api.h exists in both this PR and #6708. Was that the intention, or is this PR only suppose to contain tests?

This PR should only contain tests, and use the header in #6708

mprse commented:
> > @scartmell-arm hal/crc_api.h exists in both this PR and #6708. Was that the intention, or is this PR only suppose to contain tests?
> 
> This PR should only contain tests, and use the header in #6708

This PR provides test and header file. The header file has been added since original header file (from PR https://github.com/ARMmbed/mbed-os/pull/6708) dose not provide requirements section in form which is common for all other HAL APIs (e.g. https://github.com/ARMmbed/mbed-os/blob/9fdb5c28d81b69d16728daa81d193cbbe8c0cf98/hal/us_ticker_api.h#L58-L90).

I allowed myself to create such requirements for testing purposes and requested @scartmell-arm for review.

So this PR adds only testing requirements to original header file.

ghost commented:
In that case, it could be added as part of this PR being merged?

adbridge commented:
@mprse Is this going to need a rebase then once the other PR gets merged ? Also I'm not sure we should have test references in header files providing an API! @bulislaw can you comment please ?

mprse commented:
> Is this going to need a rebase then once the other PR gets merged ?

Probably yes.

mprse commented:
I have synced this PR with header file from #6708 yesterday. 
But the requirements should be defined in #6708 and this PR should contains only test and test header.

I can remove this commit... And we can add them later.

cmonr commented:
Awaiting the rebase!

Two things. 1) Generally, tests can come in at any time. It would have probably been better to bring this in with the other PR as a single commit, but that's alright.
2) Talked with @bulislaw about the test headers being in HAL. That's apparently alright.

adbridge commented:
> 2) Talked with @bulislaw about the test headers being in HAL. That's apparently alright.

OK not sure I totally agree with this but he is tech lead for Hal so his call.

mprse commented:
Rebased after #6708 has been merged into the master.

cmonr commented:
/morph build

cmonr commented:
Taking this outside of 5.9. Build failures indicate IAR linker issues with the new test.

Tests can come in during patch releases.

0xc0170 commented:
@mprse As the CRC was integrated, can you review the build failures?

cmonr commented:
@mprse Ping

mprse commented:
> @mprse Ping

I was out of office and I returned today.
Looks like this test is too big (NUCLEO_F070RB, NUCLEO_F072RB/IAR - out of memory). I'll try to reduce its size or split it into two tests.

mprse committed:
HAL CRC test: Fix out of memory issue on NULCLEO_F070RB, NULCLEO_F072RB/IAR

Allocate test case array on stack since memory limits on some boards.

mprse committed:
STM CRC driver: hal_crc_compute_partial() - validate input paramaters.

mprse committed:
K64F CRC driver: hal_crc_is_supported() - fix condition which validates CRC width

It looks like the intention was to return false when CRC width is different than: {16, 32} bits.

mprse committed:
K64F CRC driver: Fix handling of CRC final XOR value

According to the test results final XOR was incorrectly handled by the CRC driver.
This patch fixes this issue.

mprse committed:
K64F CRC driver: Fix coding style

mprse commented:
Provided fix for NUCLEO_F070RB, NUCLEO_F072RB/IAR - out of memory issue.
Also provided some fixes/modifications in NUCLEO and K64F CRC drivers since the tests have detected some issues. @scartmell-arm please review.

cmonr commented:
/morph build

cmonr reviewed (approved):
LGTM
